src/Groups/matrices/matrix_manipulation.jl

```diff
@@ -0,0 +1,261 @@
+      pos_i += nrows(V[k])
+      pos_j += ncols(V[k])
+   end
+   return B
+end
+
+diagonal_join(V::T...) where T <: MatElem = diagonal_join(collect(V))
+
+#=
+diagonal_join(V::T...) where T <: MatElem = cat(V; dims=(1,2))
+=#
+
+"""
+    block_matrix(m::Int, n::Int, V::AbstractVector{T}) where T <: MatElem
+Return the matrix constructed from the given block matrices, which should be given as a sequence `V` of `m*n` matrices (given in row major order, in other words listed across rows). 
+"""
+function block_matrix(m::Int, n::Int, V::AbstractVector{T}) where T <: MatElem
+   length(V)==m*n || throw(ArgumentError("Wrong number of inserted blocks"))
+   n_rows=0
+   for i in 1:m
+      for j in 1:n
+         nrows(V[n*(i-1)+j])==nrows(V[n*(i-1)+1]) || throw(ArgumentError("Invalid matrix dimension"))
+         ncols(V[n*(i-1)+j])==ncols(V[j]) || throw(ArgumentError("Invalid matrix dimension"))
+      end
+      n_rows += nrows(V[n*(i-1)+1])
+   end
+   n_cols = sum([ncols(V[j]) for j in 1:n])
+   B = zero_matrix(base_ring(V[1]), n_rows, n_cols)
+   pos_i=1
+   for i in 1:m
+      pos_j=1
+      for j in 1:n
+         insert_block!(B,V[n*(i-1)+j],pos_i,pos_j)
+         pos_j += ncols(V[n*(i-1)+j])
+      end
+      pos_i += nrows(V[n*(i-1)+1])
+   end
+   return B
+end
+
+"""
+    matrix(A::Array{AbstractAlgebra.Generic.FreeModuleElem{T},1})
+Return the matrix whose rows are the vectors in `A`. Of course, vectors in `A` must have the same length and the same base ring.
+"""
+function matrix(A::Array{AbstractAlgebra.Generic.FreeModuleElem{T},1}) where T <: FieldElem
+   c = length(A[1].v)
+   X = zero_matrix(base_ring(A[1]), length(A), c)
+   for i in 1:length(A)
+      @assert length(A[i].v)==c "Vectors must have the same length"
+      for j in 1:c X[i,j] = A[i][j] end
+   end
+
+   return X
+end
+
+"""
+    conjugate_transpose(x::MatElem{T}) where T <: FieldElem
+If the base ring of `x` is `GF(q^2)`, return the matrix `transpose( map ( y -> y^q, x) )`. An error is returned if the base ring has no even degree.
+"""
+function conjugate_transpose(x::MatElem{T}) where T <: FieldElem
+   iseven(degree(base_ring(x))) || throw(ArgumentError("The base ring must have even degree"))
+   e = div(degree(base_ring(x)),2)
+   return transpose(map(y -> frobenius(y,e),x))
+end
+
+
+# computes a complement for W in V (i.e. a subspace U of V such that V is direct sum of U and W)
+"""
+    complement(V::AbstractAlgebra.Generic.FreeModule{T}, W::AbstractAlgebra.Generic.Submodule{T})
+Return a complement for `W` in `V`, i.e. a subspace `U` of `V` such that `V` is direct sum of `U` and `W`.
+"""
+function complement(V::AbstractAlgebra.Generic.FreeModule{T}, W::AbstractAlgebra.Generic.Submodule{T}) where T <: FieldElem
+   @assert issubmodule(V,W) "The second argument is not a subspace of the first one"
+   if dim(W)==0 return sub(V,basis(V)) end
+
+   e = W.map
+
+   H = matrix( vcat([e(g) for g in gens(W)], [zero(V) for i in 1:(dim(V)-dim(W)) ]) )
+   d = dim(W)
```

Comment:
you seem to define `d` but then don't use it (or at least use `dim(W)` a lot anyway)

Reply:
I just removed it, since I'm using `dim(W)`.